File: WeeklyContests/WeeklyContests430/q3.cs
/* Could not solve it entirely during the time of the contest,
 * it has been solved post the time limit of the contest.
 */

 public class Solution
{
    private int[,] cnt = new int[1001, 1001];

    private int Gcd(int a, int b)
    {
        return b == 0 ? a : Gcd(b, a % b);
    }

    public long NumberOfSubsequences(int[] n)
    {
        long res = 0;
        Array.Clear(cnt, 0, cnt.Length);
        for (int q = n.Length - 5; q > 1; --q)
        {
            for (int s = q + 4, r = q + 2; s < n.Length; ++s)
            {
                int f = Gcd(n[r], n[s]);
                ++cnt[n[r] / f, n[s] / f];
            }
            for (int p = 0; p + 1 < q; ++p)
            {
                int f = Gcd(n[p], n[q]);
                res += cnt[n[q] / f, n[p] / f];
            }
        }
        return res;
    }
}

// The provided code utilizes several important Data Structures and Algorithms (DSA) concepts, including arrays, 
// greatest common divisor (GCD) computation, and nested loops for iteration. The code defines a class Solution with 
// a method NumberOfSubsequences that counts specific subsequences in an array of integers. It employs a 2D array cnt 
// to keep track of counts of pairs of numbers normalized by their GCD. The Gcd method is a recursive function to 
// calculate the greatest common divisor of two numbers using the Euclidean algorithm. The NumberOfSubsequences method 
// iterates through the array in a nested manner, updating the count array and using it to calculate the result based 
// on the previously stored counts. The main logic involves counting pairs of values divided by their GCD and using 
// this information to determine the number of specific subsequences that meet the criteria defined in the problem.

File: WeeklyContests/WeeklyContests434/q3.go
// Did not solve during contest time, has been solved post the contest time
//func maxFrequency(nums []int, k int) int {
// n := len(nums)
// s := make([]int, n)
// copy(s, nums)
// sort.Ints(s)

// l := 0
// t := 0
// m := 0
// for r := 0; r < n; r++ {
//     t += s[r]
//     for l <= r && s[r]*(r-l+1) - t > k {
//         t -= s[l]
//         l++
//     }
//     if r-l+1 > m {
//         m = r-l+1
//     }
// }
// return m

/*
   int res = 0;
       for (int i = 1; i <= 50; i++) {
           int cur = 0;
           for (const auto& x : nums) {
               cur = max(0, cur + (x == i) - (x == k));
               res = max(res, cur);
           }
       }
       return res + count(nums.begin(), nums.end(), k);
*/

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxFrequency(nums []int, k int) int {
	n := len(nums)
	kCount := 0
	for _, x := range nums {
		if x == k {
			kCount++
		}
	}
	answer := kCount
	for x := 1; x <= 50; x++ {
		currK := 0
		myMin := 0
		xCount := 0
		for i := 0; i < n; i++ {
			if nums[i] == x {
				xCount++
			}
			if nums[i] == k {
				currK++
			}
			myMin = min(myMin, xCount-currK)
			answer = max(answer, xCount-myMin+kCount-currK)
		}
	}
	return answer
}

/*
The provided program leverages basic data structures and algorithms to determine the
maximum frequency of any integer in the nums array, considering the effect of a specific
integer k. The max function is a utility to find the maximum of two integers, while the count
function counts occurrences of integer k in the array. The core function, maxFrequency,
iterates through potential integers (from 1 to 50), and for each integer, it traverses the
nums array to calculate a frequency-like score that considers occurrences of the current
integer and adjusts based on the presence of k. The algorithm uses nested loops to scan and
update a running count (cur) which is then compared to find the maximum frequency (res). The
final result is derived by adding the occurrences of k to the maximal frequency count. This
approach utilizes simple iteration and comparison operations to solve the problem efficiently
within the given constraints.
*/